source/assets/scripts/backend.js

```diff
@@ -0,0 +1,344 @@
+ * @param {Array<string>} intolerances - a list of intolerances. see below for details
+ * @returns {Promise} - a list of unfiltered recipe, empty if non found
+ */
+async function fetch_recipe_raw(name, intolerances) {
+  const intolerances_str = Array.isArray(intolerances) ? `&intolerances=${intolerances.join(',')}` : '';
+  const url = `https://api.spoonacular.com/recipes/complexSearch?query=${name}${intolerances_str}&apiKey=${API_KEY}&addRecipeInformation=true&fillIngredients=true`;
+  const response = await fetch(url);
+  const data = await response.json();
+  if (data.results) return data.results;
+  else return [];  // an empty list
+}
+
+/**
+ * search a recipe by its name and keep only the parts we need according to the global variable keep_fields
+ * we then remap the field names by the rule fields_remap
+ * afterwards, it saves all filtered recipes to the localstore
+ * @param {string} name - name of the recipe 
+ * @param {Array<string>} intolerances - a list of intolerances. for a list of supported keywords, see 
+ *  https://spoonacular.com/food-api/docs#:~:text=of%20supported%20diets.-,intolerances,-string
+ *  Note that this functionality is pretty bad since if a recipe can be nut free without listing nuts as intolerance
+ * @return {Promise} returns a list of fetched recipe for testing purpose
+ */
+export async function fetch_recipe(name, intolerances) {
+  const raw_recipes = await fetch_recipe_raw(name, intolerances);
+  // for each recipe, keep only ones in keep_fields and rename them accordingly
+  return raw_recipes.map(raw_recipe => {
+    const recipe = Object.create(recipe_data);  // prototype inherit recipe_data
+    // grab data from raw json
+    Object.keys(keep_fields).forEach(key => recipe[keep_fields[key]] = raw_recipe[key]);
+    recipe.steps = recipe.steps[0].steps;  // special modification: spoonacular's step array is cursed
+    return add_recipe(recipe);
+  });
+}
+
+/**
+ * compute the hash of a recipe by hashing its steps as a string
+ * @param {Object} recipe - a recipe object whose name and steps are populated
+ * @return {string} - hash of its steps as a string
+ */
+function compute_hash(recipe) {
+  if (!recipe.name) throw 'recipe has no name';
+  else if (!recipe.steps) throw 'recipe has no steps';
+
+  // https://stackoverflow.com/questions/7616461/generate-a-hash-from-string-in-javascript/7616484#7616484
+  const cyrb53 = function(str, seed = 0) {
+    let h1 = 0xdeadbeef ^ seed, h2 = 0x41c6ce57 ^ seed;
+    for (let i = 0, ch; i < str.length; i++) {
+      ch = str.charCodeAt(i);
+      h1 = Math.imul(h1 ^ ch, 2654435761);
+      h2 = Math.imul(h2 ^ ch, 1597334677);
+    }
+    h1 = Math.imul(h1 ^ (h1>>>16), 2246822507) ^ Math.imul(h2 ^ (h2>>>13), 3266489909);
+    h2 = Math.imul(h2 ^ (h2>>>16), 2246822507) ^ Math.imul(h1 ^ (h1>>>13), 3266489909);
+    return (4294967296 * (2097151 & h2) + (h1>>>0)).toString();
+  };
+  
+  return recipe.name + '$' + cyrb53(JSON.stringify(recipe.steps));
```

Comment:
note that there is a `$` between the name and the cyrb hash.
Ex. apple pie$12987198327